### PR TITLE
Displays change complexity link for female prisoners

### PR DESCRIPTION
**Pull request by Katy600**
[https://dsdmoj.atlassian.net/browse/MO-424](MO-424)

This PR  allows the HOMD to change the complexity of need level (CNL) from the HOMD allocation information pages - when you click ‘change’.

Screenshot 2021-03-26 at 19 35 56

It is then directed to an 'update complexity of need level' page

Screenshot 2021-03-26 at 19 39 32

This page has the current complexity level radio button checked. 

Once you have selected a complexity level and filled in the text area with a reason it goes to the confirmation page. 
This will display different information dependent on whether the complexity level has lowered from high or risen to high.

Screenshot 2021-03-26 at 19 45 49

If the level hasn't change then it redirects to the prisoner profile page.

### CODE REVIEW

File: app/models/complexity.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Complexity
+  include ActiveModel::Model
+  include ActiveModel::Attributes
+
+  attribute :level, :string
+  attribute :reason, :string
+
+  validates_presence_of :level
+  validates_length_of :reason, minimum: 3, message: 'Enter the reason why the level has changed'
```
Comment (starswan) [inline app/models/complexity.rb:9]: The 3 character minimum still feels abitrary and out of context with the rest of the service, which simply has presence and a maximum of 175 as its boundaries for text boxes with reasons

File: app/views/allocations/_female_complexity_level.html.erb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<tr class="govuk-table__row">
+  <td class="govuk-table__cell">Complexity level</td>
+  <td id=complexity-level class="govuk-table__cell table_cell__left_align">
+    <%= @prisoner.complexity_level.capitalize %>
+    <% unless auto_delius_import_enabled?(@prison.code) %>
+      <a class="govuk-link pull-right" href="<%= edit_prison_prisoner_complexity_level_path(@prison.code, @prisoner.offender_no) %>">Change</a>
+    <% end %>
```
Comment (starswan) [inline app/views/allocations/_female_complexity_level.html.erb:7]: This flag is a piece of legacy that we should be ripping out. There's no reason that I ca think of why we should not display this change link if this switch is set.

File: app/views/complexity_levels/confirm_complexity_changed.html.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      New complexity of need level
+    </h1>
+    <p>You have updated the complexity of need level for <%=@prisoner.first_name %> <%=@prisoner.last_name%> to:</p>
+    <%=print("COMPLEXITY LEVEL: ", @complexity) %>
+    <p id='complexity-level-update', class="govuk-body govuk-!-font-size-24 govuk-body govuk-!-font-weight-bold"><%=@complexity.level.titleize %></p>
+    <h2 class="govuk-heading-m">
+      Next steps
+    </h2>
+    <ul class="govuk-list govuk-list--bullet">
+      <%=print("COMPLEXITY LEVEL: ", @complexity) %>
```
Comment (starswan) [inline app/views/complexity_levels/confirm_complexity_changed.html.erb:14]: please remove print statement

File: app/views/complexity_levels/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <h1 class="govuk-heading-l">
+      Update complexity of need level
+    </h1>
+    <p>Complexity of need level can change during a sentence.</p>
+    <p>See EQUIP guidance on <a class="govuk-link" href="https://equip-portal.rocstac.com/CtrlWebIsapi.dll/?__id=webDiagram.show&map=0%3A9A63E167DE4B400EA07F81A9271E1944&dgm=4F984B45CBC447B1A304B2FFECABB777">how to manually reassess a person's complexity of need</a>.</p>
```
Comment (starswan) [inline app/views/complexity_levels/edit.html.erb:7]: wow does this work? We were discussing this and we didn't think we had the guidance for omplexity level on Equip yet

File: app/views/complexity_levels/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+      <%= form.govuk_collection_radio_buttons(
+            :level,
+            [OpenStruct.new(name:'High', value: 'high'),
+             OpenStruct.new(name: 'Medium', value: 'medium' ),
+             OpenStruct.new(name: 'Low', value: 'low')],
+            :value,
+            :name,
+            legend: { text: 'Select the new complexity of need level
+    '.html_safe, size: 's' }
```
Comment (starswan) [inline app/views/complexity_levels/edit.html.erb:23]: this 'html_safe' call is redundant as the text contains no markup

File: app/views/complexity_levels/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+          ) %>
+
+      <%=form.govuk_text_area(
+           :reason,
+           label: { text: 'Provide the reason(s) for this change'.html_safe, size: 's' },
```
Comment (starswan) [inline app/views/complexity_levels/edit.html.erb:28]: again html_safe call is redundant.

File: app/views/complexity_levels/edit.html.erb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+           hint: {text: 'You must justify why you are changing the complexity of need level – include any key dates and details.'}
+         ) %>
```
Comment (starswan) [inline app/views/complexity_levels/edit.html.erb:30]: I thought we normally specified 3 lines for text areas?

File: app/views/womens_allocations/show.html.erb [hunk 8]
```diff
@@ -0,0 +1 @@
+<h1>Hello</h1>
```
Comment (starswan) [inline app/views/womens_allocations/show.html.erb:1]: oops! What is meant to be in this template?

File: spec/controllers/complexity_levels_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    context 'when complexity level increases from medium to high' do
+      let(:updated_complexity_level) { 'high' }
+
+      it 'saves the complexity level and renders the renders the confirmation page' do
+        post :update,  params: { complexity: { level: updated_complexity_level, reason: 'Just because' }, prison_id: womens_prison.code, prisoner_id: offender_no }
+        expect(assigns(:previous_complexity_level)).to eq('medium')
+        expect(assigns(:complexity).level).to eq('high')
+        expect(assigns(:offender_id)).to eq(offender_no)
+        expect(assigns(:prisoner).first_name).to eq('Sally')
+        expect(assigns(:prisoner).last_name).to eq('Albright')
```
Comment (starswan) [inline spec/controllers/complexity_levels_controller_spec.rb:49]: given assertion below, maybe the target should be specified here? Maybe assert that the response is 200?

File: spec/features/complexity_level_feature_spec.rb [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+
+  before do
+    create(:allocation, nomis_offender_id: offender.fetch(:offenderNo), primary_pom_nomis_id: pom.staff_id,  prison: womens_prison.code)
+    create(:case_information, nomis_offender_id: offender.fetch(:offenderNo))
+    expect(HmppsApi::ComplexityApi).to receive(:save).with(offender.fetch(:offenderNo), level: 'low', username: "MOIC_POM", reason: 'bla bla bla')
+
+    stub_offenders_for_prison(womens_prison.code, offenders)
+    stub_signin_spo(spo, [womens_prison.code])
+    stub_offender(offender)
```
Comment (starswan) [inline spec/features/complexity_level_feature_spec.rb:20]: this line isn't needed any more IIRC

File: spec/features/complexity_level_feature_spec.rb [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    # Taken to the edit page to update the complexity level
+    expect(page).to have_text 'Update complexity of need level'
+
+    # The radio buttons should have the current complexity level checked
+    high_radio_button = find('#complexity-level-high-field', visible: false)
+    expect(high_radio_button).to be_checked
+
+    # Updated level radio button checked
+    find('label[for=complexity-level-low-field]').click
+    find('#update', visible: false)
```
Comment (starswan) [inline spec/features/complexity_level_feature_spec.rb:51]: this second find looks superfluous

File: spec/features/complexity_level_feature_spec.rb [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+    # Text area left blank
+    click_on('Update')
+    # Shows errors
+    expect(page).to have_css('.govuk-error-message')
+    expect(page).to have_text('Enter the reason why the level has changed')
+
+    # Happy path - Textarea Filled in and complexity updated from high to low
+    # find('label[for=complexity-level-low-field]').click
```
Comment (starswan) [inline spec/features/complexity_level_feature_spec.rb:60]: commented-out line of code

File: spec/features/complexity_level_feature_spec.rb [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    find('textarea#complexity-reason-field-error', visible: false)
```
Comment (starswan) [inline spec/features/complexity_level_feature_spec.rb:61]: This find doesn't seem to be required

File: spec/features/complexity_level_feature_spec.rb [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    fill_in id: 'complexity-reason-field-error', with: 'bla bla bla'
+    find('#update', visible: false)
```
Comment (starswan) [inline spec/features/complexity_level_feature_spec.rb:63]: as above - if we're waiting for the page a load, a comment here?

File: app/views/complexity_levels/edit.html.erb (hunk 4 above)
Comment (Katy600) [reply app/views/complexity_levels/edit.html.erb:7]: Oh I don't know! That is a good question!

File: app/views/complexity_levels/confirm_complexity_changed.html.erb (hunk 3 above)
Comment (Katy600) [reply app/views/complexity_levels/confirm_complexity_changed.html.erb:14]: Oh dear! Thanks.

File: app/models/complexity.rb (hunk 1 above)
Comment (Katy600) [reply app/models/complexity.rb:9]: Eve has said she doesn't want a maximum but I will change minimum 3 to presence.

File: app/views/complexity_levels/edit.html.erb (hunk 6 above)
Comment (Katy600) [reply app/views/complexity_levels/edit.html.erb:28]: Thanks

File: app/models/complexity.rb (hunk 1 above)
Comment (starswan) [reply app/models/complexity.rb:9]: In that case we need a definite comment to say that we are going against convention with our other text fields

File: spec/controllers/complexity_levels_controller_spec.rb (hunk 9 above)
Comment (Katy600) [reply spec/controllers/complexity_levels_controller_spec.rb:49]: Yes good point.

File: app/views/womens_allocations/show.html.erb (hunk 8 above)
Comment (Katy600) [reply app/views/womens_allocations/show.html.erb:1]: Oh dear

File: spec/features/complexity_level_feature_spec.rb (hunk 13 above)
Comment (Katy600) [reply spec/features/complexity_level_feature_spec.rb:61]: This one fails if I take it out, as later on, when I check to see that the page displays 'Low' in the confirmation page.

**Review by ollietreend — APPROVED** [review]
This looks excellent @Katy600 – really nice work!

I've just commented on one tiny bit of invalid HTML that I found – but that's only a small niggle.

**Pull request merged**